client/asset/eth/eth.go

```diff
@@ -313,28 +326,149 @@ func (*ExchangeWallet) PreRedeem(req *asset.PreRedeemForm) (*asset.PreRedeem, er
 	return nil, asset.ErrNotImplemented
 }
 
+// EthCoin implements the asset.Coin interface for ETH
+type EthCoin struct {
+	address common.Address
+	value   uint64
+}
+
+// address + value(uint64)
+const ethCoinIdLength = common.AddressLength + 8
+
+// ID is the ETH coins ID. It includes the address the coins came from (20 bytes)
+// and the value of the coin (8 bytes).
+func (c *EthCoin) ID() dex.Bytes {
+	id := make([]byte, ethCoinIdLength)
+	copy(id[:common.AddressLength], c.address.Bytes())
+	binary.BigEndian.PutUint64(id[common.AddressLength:], c.value)
+	return id
+}
```

Comment:
> The only other issue I see is that there are likely to be a lot of duplicate coin IDs, because two orders of the same size will have the same values = coin IDs. We are [signing the coin IDs](https://github.com/decred/dcrdex/blob/8bd3a34054f6961493cae64e8958a7e7d90e06d9/client/core/core.go#L6039) as part of order validation, and they should really be unique.

That's an interesting point.  But with UTXO assets, an order can be funded with coin A, canceled, and another order funded with coin B again.  Did you identify a bigger issue I wonder?

Reply:
> the address is accessible, and already part of the `msgjson.Trade`, so even if we returned no coins, the server would still have what it needs

Sorry, I haven't thought about `msgjson.Trade` recently so I could have this backwards, but if you were selling ETH for BTC, your `msgson.Trade.Address` would be a BTC address (from the "to" wallet), while the Coins would be for ETH (from the "from" wallet), right?

Reply:
> > the address is accessible, and already part of the `msgjson.Trade`, so even if we returned no coins, the server would still have what it needs
> 
> Sorry, I haven't thought about `msgjson.Trade` recently so I could have this backwards, but if you were selling ETH for BTC, your `msgson.Trade.Address` would be a BTC address (from the "to" wallet), while the Coins would be for ETH (from the "from" wallet), right?

Oh, yeah. Good point.

Reply:
> > The only other issue I see is that there are likely to be a lot of duplicate coin IDs, because two orders of the same size will have the same values = coin IDs. We are [signing the coin IDs](https://github.com/decred/dcrdex/blob/8bd3a34054f6961493cae64e8958a7e7d90e06d9/client/core/core.go#L6039) as part of order validation, and they should really be unique.
> 
> That's an interesting point. But with UTXO assets, an order X can be funded with coin A, order X canceled, and another order Y funded with the same coin A again. Did you identify a bigger issue I wonder?

Nothing serious. We are keying on coin ID in e.g. `trackedTrade.coins`, but as long as we only return a single coin per order, that shouldn't be a problem. `AssetCoinLocker.lockedCoins` is another place where coin ID is used as a unique key, though tons of changes are needed there regardless.

Reply:
Could also us the order ID? Hadn't thought about needing unique coin ID`s.

Reply:
Can't use order ID, since coin IDs are part of the order ID preimage.

Reply:
I really need to think about the nonce.  Having the same data represented with different (random) bytes is going to be a PITA and confusing in a host of other ways I'm almost certain.

Reply:
I haven't thought it through.. we can run with what you have

Reply:
I don't think a timestamp would be good, because possible to get two of the same.

Reply:
I like the nonce.